File: Reverse Words in a String/answer2.cpp
// a more straightforward implementation
// this code does not involve a sscanf 

class Solution {
public:
    string reverseWords(string s) {
        if(s.empty())
            return s;
        string ret;
        reverse(s.begin(),s.end());
        int i=0, j=0;
        while(j<s.size()&&s[j]==' ')   
        {
            j++;
        }
        if(j==s.size())
            return "";
        i=j;
        while(j<s.size())
        {
            i = j;
            while(j<s.size()&&s[j]!=' ')
                j++;
            reverse(s.begin()+i, s.begin()+j);
            // cout<<i<<j<<s<<endl;
            ret+=s.substr(i,j-i)+' ';
            while(j<s.size()&&s[j]==' ')   
            {
                j++;
            }    
        }
        // pop back the last space if necessary
        if(!ret.empty())
            ret.pop_back();
        // reverse(s.begin()+i, s.begin()+j);
        return ret;
    }
};

File: Reverse Words in a String/answer2.java
// a java version of answer2.cpp
// we didn't use reverser function, since I do not know if there is an equavalent in Java

class Solution {
    public String reverseWords(String s) {
        if(s.isEmpty())
            return s;
        
        
        // traverse the string from the last element
        int i=s.length()-1, j=s.length()-1;
        while(j>=0&&s.charAt(j)==' ')   
        {
            j--;
        }
        if(j<0)
            return "";
        i=j;
        
        // unlike C++, we need to initialize the String.
        // even though it is only an empty string
        String ret = "";
        while(j>=0)
        {
            i = j;
            while(j>=0&&s.charAt(j)!=' ')
                j--;
            ret+=s.substring(j+1,i+1)+' ';
            while(j>=0&&s.charAt(j)==' ')   
            {
                j--;
            }    
        }
        // pop_back() does not have an equavalent in Java
        // thus, I use a substring function of extract the string we need
        if(!ret.isEmpty())
            return ret.substring(0,ret.length()-1);
        else
            return "";
    }
}
